src/ripple/shamap/README.md

```diff
@@ -26,164 +37,307 @@ There are two different ways of building and using a SHAMap:
+the constructor called by `make_shared` throws we don't want to break our
+promise.
 
 
 ## Canonicalize ##
 
-The calls to `canonicalize()` make sure that if the resulting node is already
-in the SHAMap, then we return the node that's already present -- we never
-replace a pre-existing node.  By using `canonicalize()` we manage a thread
-race condition where two different threads might both recognize the lack of a
-SHAMapTreeNode at the same time.  If they both attempt to insert the node
-then `canonicalize` makes sure that the first node in wins and the slower
-thread receives back a pointer to the node inserted by the faster thread.
+A call to `canonicalize()` stores the node in the cache if it does not already
+exist in the cache.
+
+The calls to `canonicalize()` make sure that if the resulting node is already in
+the SHAMap, node cache or database, then we don't create duplicates.
+
+On retrieving existing nodes, the copy in the database has top priority.  If
+there's a duplicate in the cache, it is replaced by the copy in the database. Or
+if there is a duplicate found between the cache and the SHAMap, the copy in the
+cache prevails and replaces that in the SHAMap.
+
+On inserting new nodes (such as new transactions or account states), the
+opposite priority holds.  The SHAMap dominates copies in the cache and database
+(if found), and the cache dominates any copies that may be found in the
+database.
+
+By using `canonicalize()` we manage a thread race condition where two different
+threads might both recognize the lack of a SHAMapTreeNode at the same time
+(during a fetch).  If they both attempt to insert the node into the SHAMap, then
+`canonicalize` makes sure that the first node in wins and the slower thread
+receives back a pointer to the node inserted by the faster thread.
+
+Conversely, two different trees can have the exact same node (the same hash
+value) with two different IDs.  If the older tree caches the node, and then the
+node is inserted into a new tree (e.g. a transaction that didn't make the last
+ledger), then that transaction gets a new id and dominates the transaction node
+in the cache.
+
+The TreeNodeCache is not subject to the rule that any node must be resident
+forever.  So it's okay to replace the old node with the new node.
+
+## TreeNodeCache ##
+
+The `TreeNodeCache` is a `std::unordered_map` keyed on the hash of the
+SHAMap node.  The stored type consists of `shared_ptr<SHAMapAbstractNode>`,
+`weak_ptr<SHAMapAbstractNode>`, and a time point indicating the most recent
+access of this node in the cache.  The time point is based on
+`std::chrono::steady_clock`.
+
+The container uses a cryptographically secure hash that is randomly seeded.
+
+The `TreeNodeCache` also carries with it various data used for statistics
+and logging, and a target age for the contained nodes.  When the target age
+for a node is exceeded, and there are no more references to the node, the
+node is removed from the `TreeNodeCache`.
```

Comment:
Is TreeNodeCache the cache discussed in section Canonicalize? Can we discuss more on when TreeNodeCache is needed and how it is being used?

Reply:
Yes, it is the TreeNodeCache.  I've replaced uses of the word cache with TreeNodeCache in this section to clarify that point.  

The Canonicalize section is the best place to describe how the TreeNodeCache is used as that is the only place to use it.  I've added to that a description of when `canonicalize` is called.